### ISSUE TITLE
Simultaneous connections error timeout Windows Server 2008

### ISSUE THREAD

**Issue by fbacker** [socketio/socket.io#2338]
I'm using Socket.IO on Windows Server 2008. It is clustered to use all CPUs and will use multiple servers as well so I'm using Redis to allow clusters to talk to each other.

However when one cluster gets over 4000 connections I'm starting to get error timeout on the clients.

    socket.on('connect_error', function(error){
        console.log("conn error "+ error);
    });

This is the server code
https://github.com/fbacker/NodeChat-SocketIO/blob/master/app.js

This issue also occurs when running on AWS Linux instances with ulimit set to 65000

**Comment by fbacker** [comment]
Found the issue. Socket.io cannot handle more than 2.500 connections for each CPU (cluster).

**Comment by LordMajestros** [comment]
Really? Is this in the documentation?

**Comment by fbacker** [comment]
Guessing that if you doing a ping/pong test you can get more connections but when having some logic I really have trouble getting many connections. Also sticky-sessions based on IP isn't that good of an solution either.

Am in a critical project with live tv in a couple of days so can't change anything so I've now setup 24 single CPU servers on AWS to get 50.000 users.
Else I'm using the websocket really wrong.

However I'll try to create my own http-proxy with some logic not to overflow a single cluster with to many connections.

Interested, here's my code for the moment. Will update it hopefully after new years.
https://github.com/fbacker/NodeChat-SocketIO

**Comment by LordMajestros** [comment]
I am aware of open files/socket, memory and processor limits. I reached 4000 on a test across a LAN to a system with 2 CPUs I can't recall right now if I used cluster to use both CPUs but I was doing a broadcast for live prices every 45 seconds to each of the 4000 clients with no lag or loss. This was on a LAN so it probably won't translate to real world usage. I will try your code and post results.

**Comment by LordMajestros** [comment]
I achieved >4000 concurrent connections on a single core/cpu without using cluster.js
![above 4000](https://cloud.githubusercontent.com/assets/10727946/11912827/565e19be-a64f-11e5-8f48-a70823944218.png)
Using cluster.js on 2 cores/cpus I achieved >6000 with >2500 per core
Server 
![6000 clients server](https://cloud.githubusercontent.com/assets/10727946/11912829/876bd67c-a64f-11e5-8840-28a85375b51e.png)
Load generator 
![6000 clients on 2 nodes](https://cloud.githubusercontent.com/assets/10727946/11912830/ad12b5ee-a64f-11e5-838a-ae2f1fbff2a2.png)

Each client connects via regular http then on upgrade to web socket a disconnect is fired then a reconnecting while it upgrades and then connected successfully on successful upgrade to websocket.

I don't think there is 2500 limit per core in cluster.
This was run on node 0.12.7, npm 2.11.3 ubuntu 15.04(vivid vervet) 4gb ram after raising ulimit and other settings.

I also achieved similar results on node 4.2.1, npm 2.14.7 on OS X yosemite 16 gb ram

Ram use in both cases was ~ 300mb and no broadcast messages from server to all clients were dropped

I could have used higher concurrencies but I only set out to see if there was a 2500 limit per core/cpu in cluster.js

**Comment by LordMajestros** [comment]
This is the testing code I used
```
/**
 * Concurrency test for socket.io realtime engine
 * @author Dumbi Olise
 * @date Sat Nov 21, 2015 (21:37:54)
 */
 // SOCKET IO bench
var io = require('socket.io-client'),
chalk = require('chalk'),
opts = {
  transports: ['websocket'],//'polling'],
  'force new connection': true, // undocumented option
  reconnectionAttempts:7,
  reconnectionDelayMax: 500,
  reconnectionDelay:50,
  path:'/b'
},
connectCount = 0,
disconnectCount = 0,
updateMessageCount = 0,
priceDropMessageCount = 0,
tempDisconnect = 0,
m_MessageCount = 0,
reconnectCount = 0,
showDisconnects = true;
setEvents=function(client,index){
	var onConnect=function(){
		connectCount++;
		if(showDisconnects)
			console.log('Client '+index+' connected successfully');
	},
	onReconnectAttempt=function(clientNum){
		reconnectCount++;
		tempDisconnect++;
		if(showDisconnects)
			console.log('Client '+clientNum+' reconnecting');
	}
	onReconnectSuccessful=function(clientNum){
		connectCount++;
		if(showDisconnects)
			console.log('Client '+clientNum+' reconnected successfully');
	}
	onDisconnect=function(clientNum){
		if(showDisconnects)
			console.log('Client '+clientNum+' disconnected unexpectedly');
		disconnectCount++;
	},
	onUpdateMessage=function(){
		updateMessageCount++;
	},
	onPriceDropMessage=function(){
		priceDropMessageCount++;
	},
	on_M_Message=function(){
		m_MessageCount++;
	};
	if(client){
		client.on('connect',onConnect);
		client.on('reconnect',onReconnectSuccessful(index));
		client.on('reconnect_failed',onDisconnect(index));
		client.on('reconnect_attempt',onReconnectAttempt(index));
		client.on('u',onUpdateMessage);
		client.on('p',onPriceDropMessage);
		client.on('m',on_M_Message);
	}
	return client;
},
host = 'localhost',
port = 3000,
host_port = null,
interval =null,
slowInterval=50,
slow = 0,
load = 1000,
program = require('commander'),
clients=[];

function createClient(index, controller) {
  return setEvents(io.connect(host_port, opts),index);
}
program.version('0.1')
    .option('-p, --port [num]', 'Set the server port (default 8080)')
    .option('-h, --host [str]', 'Set the socket.io host to connect to')
    .option('-l, --load [num]', 'Set the number of clients to create (load)')
    .option('-s, --slow [num]', 'Set to >0 to increase load slowly')
    .option('-i, --slowInterval [num]', 'Set to >0 to specify slow loading interval')
    .parse(process.argv);
if(program.port) {
    port = program.port;
}
if(program.host){
	host = program.host;
}
if(program.load){
	load = program.load;
}
if(program.slow){
	slow = program.slow;
	if(program.slowInterval){
		slowInterval = program.slowInterval;
	}
}
host_port = 'http://'+host+':'+port;//http is necessary or it defults to https
console.log(chalk.blue('Connection to '+host+' on port '+port+'('+host_port+')\nLoad:'+load));
if(slow==0){
	for(var i=0;i<load;i++){
		clients.push(createClient(i));
	}
}
else{
	interval=setInterval(function(){
		clients.push(createClient(clients.length));
		if(clients.length>=load){
			clearInterval(interval);
			priceDropMessageCount = 0;
		}			
	},slowInterval);
}

function showStatus(){
	console.log(chalk.green('\n\nLoad: '+load+
		'\nConnect count: '+connectCount+
		'\nDisconnect count: '+disconnectCount+
		'\nUpdate messages: '+updateMessageCount+
		'\nPrice drop messages: '+priceDropMessageCount+
		'\nm messagecount: '+m_MessageCount+
		'\ntemporary Disconnects: '+tempDisconnect+
		'\nReconnect attempts: '+reconnectCount+
		'\nCurrently connected:'+ (connectCount-disconnectCount>0?connectCount-disconnectCount:0)));
}
setInterval(showStatus,45000);
process.on('SIGINT', function() {
	showDisconnects = false;
	process.exit(0);
});


```

**Comment by LordMajestros** [comment]
I changed the default socket.io connection path on the server to /b so I set that in the options here.
Also the server emits broadcasts every 45 seconds

**Comment by LordMajestros** [comment]
Maybe this is a windows server 2008 issue.

**Comment by fbacker** [comment]
I've been trying back and fort between windows and linux making it to work as stable as possible before the 23;rd of December. 4.000 sounds familiar, perhaps I got that as well when running on better machines and 2.500 running AWS t2.micro instances. 
Sticky-load balancer based on IP doesn't feel like a good solution so I've started 24 servers with single CPU;s to get it to work. 
Will write a custom http-proxy instead and then I can use multiple CPU;s and servers again.
